sql/core/src/main/scala/org/apache/spark/sql/ui/SQLListener.scala

```diff
@@ -0,0 +1,340 @@
+      }
+      if (executionUIData.completionTime.nonEmpty && !executionUIData.hasRunningJobs) {
+        // onExecutionEnd happens before this onJobEnd and we are the last job, so we should update
+        // the execution lists.
+        updateExecutionLists(executionId)
+      }
+    }
+  }
+
+  override def onExecutorMetricsUpdate(
+      executorMetricsUpdate: SparkListenerExecutorMetricsUpdate): Unit = synchronized {
+    for ((taskId, stageId, stageAttemptID, metrics) <- executorMetricsUpdate.taskMetrics) {
+      updateTaskMetrics(taskId, stageId, stageAttemptID, metrics, false)
+    }
+  }
+
+  override def onStageSubmitted(stageSubmitted: SparkListenerStageSubmitted): Unit = synchronized {
+    val stageId = stageSubmitted.stageInfo.stageId
+    val stageAttemptId = stageSubmitted.stageInfo.attemptId
+    // Always override metrics for old stage attempt
+    stageIdToStageMetrics(stageId) = SQLStageMetrics(stageAttemptId)
+  }
+
+  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+    updateTaskMetrics(
+      taskEnd.taskInfo.taskId, taskEnd.stageId, taskEnd.stageAttemptId, taskEnd.taskMetrics, true)
+  }
+
+  private def updateTaskMetrics(
+      taskId: Long,
+      stageId: Int,
+      stageAttemptID: Int,
+      metrics: TaskMetrics,
+      finishTask: Boolean): Unit = {
+    if (metrics == null) {
+      return
+    }
+
+    stageIdToStageMetrics.get(stageId) match {
+      case Some(stageMetrics) =>
+        if (stageAttemptID < stageMetrics.stageAttemptId) {
+          // A task of an old stage attempt. Because a new stage is submitted, we can ignore it.
+        } else if (stageAttemptID > stageMetrics.stageAttemptId) {
+          // TODO A running task with a higher stageAttemptID??
+        } else {
+          // TODO We don't know the attemptId. Currently, what we can do is overriding the
+          // accumulator updates. However, if there are two same task are running, such as
+          // speculation, the accumulator updates will be overriding by different task attempts,
+          // the results will be weird.
+          stageMetrics.taskIdToMetricUpdates.get(taskId) match {
+            case Some(taskMetrics) =>
+              if (finishTask) {
+                taskMetrics.finished = true
+                taskMetrics.accumulatorUpdates = metrics.accumulatorUpdates()
+              } else if (!taskMetrics.finished){
+                // If a task is finished, we should not override with accumulator updates from
+                // heartbeat reports
+                taskMetrics.accumulatorUpdates = metrics.accumulatorUpdates()
+              }
+            case None =>
+              // TODO Now just set attemptId to 0. Should fix here when we can get the attempt
+              // id from SparkListenerExecutorMetricsUpdate
+              stageMetrics.taskIdToMetricUpdates(taskId) =
+                SQLTaskMetrics(attemptId = 0, finished = finishTask, metrics.accumulatorUpdates())
```

Comment:
Is the following a valid race condition:
(1) Executor sends heartbeat
(2) The task finishes, and we call `onTaskEnd`
(3) We get the heartbeat, and we call `onExecutorMetricsUpdate`

So the result is that even though the task is finished it may still be marked as incomplete. Should we just ignore all heartbeat updates if the task has already finished?